pkg/controller/jobframework/integrationmanager.go

```diff
@@ -186,6 +197,23 @@ func ForEachIntegration(f func(name string, cb IntegrationCallbacks) error) erro
 	return manager.forEach(f)
 }
 
+// EnableIntegration marks the integration identified by name as enabled.
+func EnableIntegration(name string) {
+	manager.enableIntegration(name)
+}
+
+// EnableIntegrationsForTest - should be used only in tests
+// Mark the frameworks identified by names and return a revert function.
+func EnableIntegrationsForTest(names ...string) func() {
```

Comment:
I think that we should not put any function only for testing.
Could you move the testing utils package?

I guess that implementing the `GetEnabledIntegrations` function in this package allows us to move this function to another package, right?

Reply:
Not relay we need to be able to revert the changes done. we will need to add two extra public funcs  one to get the integrations previously enabled and another one to restore the configuration, the new funcs will also only be used in testing.

Reply:
I guess an alternative can be to return the revert function from `EnableIntegration` and ignore the returned function when not in a test. (But I don't like this idea too much)

Reply:
Another possibility would be to create a new instance of integrationManager per test case.

Reply:
The code impact will be too much with a close to 0 added value.

Reply:
Adding functions, into production code, just for testing should be avoided (if feasible, and it seems that with the new instance it could work well). 

Code should generally separate concerns. In particular, separate business responsibilities and testing to make the production code focused, making it easier to understand and thus maintain.

Reply:
> Not relay we need to be able to revert the changes done. we will need to add two extra public funcs one to get the integrations previously enabled and another one to restore the configuration, the new funcs will also only be used in testing.

If so, shouldn't we receive the `integration manager` as an EnableIntegrationsForTesting argument?
I believe that `SetFeatureGateDuringTest` function would be a good example.

> Adding functions, into production code, just for testing should be avoided (if feasible, and it seems that with the new instance it could work well).
> 
> Code should generally separate concerns. In particular, separate business responsibilities and testing to make the production code focused, making it easier to understand and thus maintain.

I fully agree with you. We never want to mix the production codes and testing code.
Additionally, this function is exposed, which means the third party providers could import.

Reply:
> > Not relay we need to be able to revert the changes done. we will need to add two extra public funcs one to get the integrations previously enabled and another one to restore the configuration, the new funcs will also only be used in testing.
> 
> If so, shouldn't we receive the `integration manager` as an EnableIntegrationsForTesting argument? I believe that `SetFeatureGateDuringTest` function would be a good example.
> 

The is no knowledge of `integrationManager` type existence outside the jobframework package and in my opinion it should stay that way. For the rest of the integration manager is just a global state that being the reason why we need a revert mechanism.


> > Adding functions, into production code, just for testing should be avoided (if feasible, and it seems that with the new instance it could work well).
> > Code should generally separate concerns. In particular, separate business responsibilities and testing to make the production code focused, making it easier to understand and thus maintain.
> 
> I fully agree with you. We never want to mix the production codes and testing code. Additionally, this function is exposed, which means the third party providers could import.

Go being able to support build time tests only funcs out of the box will be ideal however we can only try to:
- Put the test specific code that needs to be accessed from outside the package in conditional built file `//go:build ....`, and only use tat build tag when running the unit test
- Panic in the test specific code if `!testing.Testing()`

However both are overkill in my opinion.

Also this in not the only `test dedicated`  implementation in kueue, https://github.com/kubernetes-sigs/kueue/blob/a3688c4a1bd30743414c3e1bef71369dfde13521/pkg/scheduler/scheduler.go#L76 is just one other example.

Reply:
I don't have a clearly advantageous approach, so I'm willing to accept it, unless maybe @alculquicondor can recommend something. In fact this issue would show up for external frameworks in https://github.com/kubernetes-sigs/kueue/pull/2059/, but if we wanted to modify its state from other unit tests.

Reply:
> Also this in not the only test dedicated implementation in kueue,
> 
> [kueue/pkg/scheduler/scheduler.go](https://github.com/kubernetes-sigs/kueue/blob/a3688c4a1bd30743414c3e1bef71369dfde13521/pkg/scheduler/scheduler.go#L76)
> 
> Line 76 in [a3688c4](https://github.com/kubernetes-sigs/kueue/commit/a3688c4a1bd30743414c3e1bef71369dfde13521)
> 
>  applyAdmission func(context.Context, *kueue.Workload) error 
> is just one other example.

Significantly differences between this one and the `applyAdmission` is if the knob is exposed to other packages.
In the case of `applyAdmission`, we can not modify the field from other packages.

So, at least, it would be more better to add a testing object as a `EnableIntegrationsForTest` arguments like this. This is generic approach to prevent accidently used by production codes. After we find better approach, we can revisit here.

```suggestion
// EnableIntegrationsForTest - should be used only in tests
// Mark the frameworks identified by names and return a revert function.
func EnableIntegrationsForTest(tb testing.TB, names ...string) {
	tb.Helper()
	[...]
	tb.Cleanup(func(
		manager.enabledIntegrations = old
	))
```

Reply:
> Significantly differences between this one and the `applyAdmission` is if the knob is exposed to other packages. In the case of `applyAdmission`, we can not modify the field from other packages.

That is irrelevant if we talk about having **only for testing** code (`I think that we should not put any function only for testing.`) and `applyAdmission` is not the only one.

> So, at least, it would be more better to add a testing object as a `EnableIntegrationsForTest` arguments like this. This is generic approach to prevent accidently used by production codes. After we find better approach, we can revisit here.

Done